src/ai/defaultai.cc

```diff
@@ -841,31 +866,141 @@ void DefaultAI::late_initialization() {
+						 *
+						 * */
+						for (const auto& attribute : prod.collected_attributes()) {
+							buildings_immovable_attributes_[attribute.second].insert(
+							   ImmovableAttribute(bo.name, BuildingAttribute::kNeedsRocks));
+						}
+					} else {
+						log_dbg_time(gametime, "AI %d detected lumberjack: %s", player_number(), bo.name);
+						bo.set_is(BuildingAttribute::kLumberjack);
+						/* Buildings detected at the time of writing:
+						 *
+						 *   amazons_rare_tree_cutters_hut NOCOM new, used to be kSupportingProducer
+						 *   amazons_woodcutters_hut
+						 *   atlanteans_woodcutters_house
+						 *   barbarians_lumberjacks_hut
+						 *   empire_lumberjacks_house
+						 *   frisians_woodcutters_house
+						 *
+						 * */
+						for (const auto& attribute : prod.collected_attributes()) {
+							buildings_immovable_attributes_[attribute.second].insert(
+							   ImmovableAttribute(bo.name, BuildingAttribute::kLumberjack));
+						}
+						lumberjacks.insert(&prod);
+					}
+				}
+				if (produces_non_construction_material) {
+					log_dbg_time(
+					   gametime, "AI %d detected berry collector: %s", player_number(), bo.name);
+					bo.set_is(BuildingAttribute::kNeedsBerry);
+					/* Buildings detected at the time of writing:
+					 *
+					 *   frisians_collectors_house
+					 *
+					 * */
+
+					for (const auto& attribute : prod.collected_attributes()) {
+						buildings_immovable_attributes_[attribute.second].insert(
+						   ImmovableAttribute(bo.name, BuildingAttribute::kNeedsBerry));
+					}
+				}
 			}
+
 			// here we identify hunters
-			if (bh.collects_ware_from_map() == "meat") {
+			if (!prod.collected_bobs().empty()) {
+				log_dbg_time(gametime, "AI %d detected hunter: %s", player_number(), bo.name);
 				bo.set_is(BuildingAttribute::kHunter);
-			}
-			// and fishers
-			if (bh.collects_ware_from_map() == "fish" && bo.inputs.empty()) {
+				/* Buildings detected at the time of writing:
+				 *
+				 *   amazons_hunter_gatherers_hut NOCOM new
+				 *   atlanteans_hunters_house
+				 *   barbarians_hunters_hut
+				 *   empire_hunters_house
+				 *   frisians_hunters_house
+				 *
+				 * */
+			}
+
+			// fishers
+			if (bh.needs_water() && prod.collected_resources().count("resource_fish") == 1) {
+				log_dbg_time(gametime, "AI %d detected fisher: %s", player_number(), bo.name);
 				bo.set_is(BuildingAttribute::kFisher);
+				/* Buildings detected at the time of writing:
+				 *
+				 *   amazons_hunter_gatherers_hut
+				 *   atlanteans_fishers_house
+				 *   barbarians_fishers_hut
+				 *   empire_fishers_house
+				 *   frisians_fishers_house
+				 *
+				 * */
 			}
-			// and collectors
-			if (bh.collects_ware_from_map() == "fruit") {
-				bo.set_is(BuildingAttribute::kNeedsBerry);
+
+			// wells
+			if (prod.input_wares().empty()) {
+				for (Widelands::DescriptionIndex ware_index : prod.output_ware_types()) {
+					if (tribe_->get_ware_descr(ware_index)->name() == "water") {
```

Comment:
this definition is now different then before as well. in trunk only buildings that mine water from the ground are defined as well now this is dependent on the output. Does not do any harm currently as the amazon water gatherer was faked to be such a well but if we can lower the restriction to have one well defined we might need to change this again

Reply:
will fix this in another branch after this goes in. together with some other optimizations especially for amazons